utils/frame/remote-externalities/src/lib.rs

```diff
@@ -341,26 +365,216 @@ impl<B: BlockT + DeserializeOwned> Builder<B> {
+	/// Save the given data to the child keys snapshot.
+	fn save_child_snapshot(
+		&self,
+		data: &ChildKeyValues,
+		path: &PathBuf,
+	) -> Result<(), &'static str> {
+		let mut path = path.clone();
+		path.set_extension("child");
+		let encoded = data.encode();
+		debug!(
+			target: LOG_TARGET,
+			"writing {} bytes to state snapshot file {:?}",
+			encoded.len(),
+			path
+		);
+		fs::write(path, encoded).map_err(|_| "fs::write failed.")?;
+		Ok(())
+	}
+
+	fn load_top_snapshot(&self, path: &PathBuf) -> Result<TopKeyValues, &'static str> {
+		let mut path = path.clone();
+		path.set_extension("top");
+		info!(target: LOG_TARGET, "loading top key-pairs from snapshot {:?}", path);
+		let bytes = fs::read(path).map_err(|_| "fs::read failed.")?;
+		Decode::decode(&mut &*bytes).map_err(|e| {
+			log::error!(target: LOG_TARGET, "{:?}", e);
+			"decode failed"
+		})
+	}
+
+	fn load_child_snapshot(&self, path: &PathBuf) -> Result<ChildKeyValues, &'static str> {
+		let mut path = path.clone();
+		path.set_extension("child");
+		info!(target: LOG_TARGET, "loading child key-pairs from snapshot {:?}", path);
 		let bytes = fs::read(path).map_err(|_| "fs::read failed.")?;
-		Decode::decode(&mut &*bytes).map_err(|_| "decode failed")
+		Decode::decode(&mut &*bytes).map_err(|e| {
+			log::error!(target: LOG_TARGET, "{:?}", e);
+			"decode failed"
+		})
+	}
+
+	/// Load all the `top` keys from the remote config, and maybe write then to cache.
+	async fn load_top_remote_and_maybe_save(&self) -> Result<TopKeyValues, &'static str> {
+		let top_kv = self.load_top_remote().await?;
+		if let Some(c) = &self.as_online().state_snapshot {
+			self.save_top_snapshot(&top_kv, &c.path)?;
+		}
+		Ok(top_kv)
+	}
+
+	/// Load all of the child keys from the remote config, given the already scraped list of top key
+	/// pairs.
+	///
+	/// Stores all values to cache as well, if provided.
+	async fn load_child_remote_and_maybe_save(
+		&self,
+		top_kv: &[KeyValue],
+	) -> Result<ChildKeyValues, &'static str> {
+		let child_kv = self.load_child_remote(&top_kv).await?;
+		if let Some(c) = &self.as_online().state_snapshot {
+			self.save_child_snapshot(&child_kv, &c.path)?;
+		}
+		Ok(child_kv)
+	}
+
+	/// Load all of the child keys from the remote config, given the already scraped list of top key
+	/// pairs.
+	///
+	/// `top_kv` need not be only child-bearing top keys. It should be all of the top keys that are
+	/// included thus far.
+	async fn load_child_remote(&self, top_kv: &[KeyValue]) -> Result<ChildKeyValues, &'static str> {
+		let child_bearing_top_keys = top_kv
+			.iter()
+			.filter_map(|(k, _)| is_default_child_storage_key(k.as_ref()).then(|| k))
+			.collect::<Vec<_>>();
+
+		info!(
+			target: LOG_TARGET,
+			"👩‍👦 scraping child-tree data from {} top keys",
+			child_bearing_top_keys.len()
+		);
+
+		let mut child_kv = vec![];
+		for prefixed_top_key in child_bearing_top_keys {
```

Comment:
I got really confused, why don't you just call the child storage roots? :D

Reply:
renamed this instance. Generally speaking, the jargon here is super confusing. Kinda hard to specify what node is in what tree. I though `child_bearing_top_keys` is pretty accurate actually.